### DISCUSSION TITLE
[Chapter 06] 6.3 Interceptor, servletFilter  & locale

### DISCUSSION THREAD

**Discussion by Youngju-Jang** [cowaine/spring-boot-for-msa#55] [Chapter 06. Building a Web Application Server] [answered]
1. Interceptor, servletFilter
   - filter 의 경우 `dispatcherServlet` 앞에서 사용, 인터셉터는 `dispatcherServlet` 이후에 실행된다고 보이는데,
   - 1) 현업에서는 주로 어떤 걸 더 많이 쓰나요? 
   - 2) `인터셉터`는 스프링 프레임워크 기반 기술, `필터`는 표준스펙 제공이라고 하는데, 
         이럴때 `필터`는 POJO에 맞고, `인터셉터`는 맞지않다고 표현하는게 맞을까요?

2. formatter, converter
   - 예제에서 `HotelRoomNumber` 의 `formatter`, `converter`가 두개 다 있는데,
   - 둘 다 만들어두고 둘 다 @component애노테이션으로 등록시켜둘 경우 어떤걸로 적용이 될까요?
   - 그리고 예제에서 `Locale`을 파라미터로 받는게 있던데, json으로 보낼때 `locale`을 따로 같이 보내주면 적용이 되는걸까요? 
   ```java
    @Override
         public HotelRoomNumber parse(String text, Locale locale) throws ParseException {
              return HotelRoomNumber.parse(text);
         }
    ```

**Comment by ahngilwoong** [comment, marked as answer]
**1-A.** 저는 입사하고 둘 다 한번도 써보지 못했지만 제 생각엔 어떤 분야의 어떤 파트를 맡게 되느냐에 따라서 달라질 것 같습니다.

**1-B.** 인터셉터가 스프링 프레임웍에 결합도가 커서 말씀하신게 맞는 것 같은데 필터도 톰캣(웹컨테이너)가 돌려주는것으로 알고 있어서 저도 다른 사람의 의견을 듣고싶습니다.

**2-1.** 저도 궁금해서 찾아보았는데 친절하게 설명된 블로그가 있어 공유드립니다.

 **스프링 MVC 컨버터와 포멧터의 우선순위**
 1. 직접 등록한 컨버터
 2. 자동 등록된 컨버터
 3. 직접 등록한 포멧터
 4. 자동 등록한 포멧터

> 컨버터와 포멧터는 유사한 영역을 바라보고 있다. 왜냐하면 포멧터가 String 타입을 기준으로 동작을 하기 때문이다. 따라서, 등록한 컨버터와 포멧터는 위의 우선순위를 고려해서 동작한다. 기본적으로 컨버터의 우선순위가 포멧터보다 높다. 
라고 합니다..
[컨버터와 포매터 우선순위 참고 블로그](https://ojt90902.tistory.com/679)

**2-2.** 책 내용에 AcceptHeaderLocaleResolver는 사용자 요청 메시지의 Accept-Language 헤더에 값을 사용하여 Locale 객체를 생성 한다, HTTP 요청 메시지의 Accept-Language 헤더 값을 파싱하여 Locale 객체로 변경한다.  라고 적힌 부분이 있는데 제가 원하는 정도로 자세히 설명이 되어있진 않더라구요 
근데 제가 이해한 책의 내용은 JSON에 값을 넣기보다는
Accept-Language에 Locale 정보를 넣거나 RequestParam 값으로 Locale 정보를 넘기자 라고 이해했습니다
값은 이렇게 넘어가는 것 같습니다. 
`Accept-Language: de-CH`
`GET /hotels?locale=ko`
아래 링크는 de-CH가 뭐지? 에서 찾은 링크입니다
[ISO 랭귀지코드 정보](http://www.lingoes.net/en/translator/langcode.htm)

**Comment by BvrPark** [comment]
길웅 선생님의 답변이 너무 완벽하여 감탄을 표하고 갑니다😲
